tests/MongoDB.Driver.Core.Tests/Core/Bindings/CoreSessionTests.cs

```diff
@@ -240,14 +243,148 @@ public void WasUsed_should_call_serverSession()
 
             Mock.Get(subject.ServerSession).Verify(m => m.WasUsed(), Times.Once);
         }
+        
+        [Theory]
+        [ParameterAttributeData]
+        public void EnsureTransactionsAreSupported_should_throw_when_there_are_no_connected_servers(
+            [Values(0, 1, 2, 3)] int numberOfDisconnectedServers)
+        {
+            var clusterDescription = CreateClusterDescriptionWithDisconnectedServers(numberOfDisconnectedServers);
+            var mockCluster = new Mock<ICluster>();
+            mockCluster.SetupGet(m => m.Description).Returns(clusterDescription);
+            var subject = CreateSubject(cluster: mockCluster.Object);
+
+            var exception = Record.Exception(() => subject.EnsureTransactionsAreSupported());
+
+            var e = exception.Should().BeOfType<NotSupportedException>().Subject;
+            e.Message.Should().Be("StartTransaction cannot determine if transactions are supported because there are no connected servers.");
+        }
+
+
+        [Theory]
+        // Scenario: 0 = Disconnected, 1 = Connected does not support transactions, 2 = Connected supports transactions
+        [InlineData(new[] { 1 })]
+        [InlineData(new[] { 0, 1 })]
+        [InlineData(new[] { 1, 0 })]
+        [InlineData(new[] { 1, 1 })]
+        [InlineData(new[] { 1, 2 })]
+        [InlineData(new[] { 2, 1 })]
+        [InlineData(new[] { 0, 0, 1 })]
+        [InlineData(new[] { 0, 1, 0 })]
+        [InlineData(new[] { 0, 1, 1 })]
+        [InlineData(new[] { 0, 1, 2 })]
+        [InlineData(new[] { 1, 1, 0 })]
+        [InlineData(new[] { 1, 1, 1 })]
+        [InlineData(new[] { 1, 1, 2 })]
+        [InlineData(new[] { 2, 1, 0 })]
+        [InlineData(new[] { 2, 1, 1 })]
+        [InlineData(new[] { 2, 1, 2 })]
+        public void EnsureTransactionsAreSupported_should_throw_when_any_connected_replica_set_primary_does_not_support_transactions(int[] scenarios)
+        {
```

Comment:
I'm using the scenario codes as a compact way to encode the various combinations of connected/disconnected and supports/does not support transactions.

Reply:
Ack